### PR TITLE
Add article about python good practices

### PR DESCRIPTION
**Pull request by leplatrem**

### CODE REVIEW

File: content/Dev/python_good_practices_2020.rst [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+
+    from poucave.app import main
+
+    main(sys.argv[1:])
+
+The most appreciated libraries for advanced CLI parameters seem to be `Click <https://click.palletsprojects.com>`_ (declarative) and `Fire <https://github.com/google/python-fire>`_ (automatic).
+
+For the Docker container, we follow our `Dockerflow conventions <https://github.com/mozilla-services/Dockerflow>`_. This helps our OPs team to treat all containers the same way, regardless of implementation language etc.
+
+A good take away for any application is to manage configuration through environment variables.
```
Comment (areski) [inline content/Dev/python_good_practices_2020.rst:216]: "any application deployment" ?

also ref to https://12factor.net/config

File: content/Dev/python_good_practices_2020.rst [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+        data = cli.get("/")
+
+        assert data["app"] == "poucave"
+
+    def test_api_fetches_info_from_source(cli, mock_aioresponses):
+        mock_aioresponses.get(config.SOURCE_URI, json={"success": True})
+
+        cli.get("/check-source")
+
+
```
Comment (areski) [inline content/Dev/python_good_practices_2020.rst:360]: Is it me or I like a conclusion :)

File: content/Dev/python_good_practices_2020.rst [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+The working combination in one ``Makefile`` target is:
+
+.. code-block:: make
+
+    format: $(INSTALL_STAMP)
+        $(VENV)/bin/isort --line-width=88 --lines-after-imports=2 -rc $(SOURCE) --virtual-env=$(VENV)
+        $(VENV)/bin/black $(SOURCE)
+
+Again, to avoid having an extra configuration file for *isort* we'll use CLI arguments :)
```
[the remark this answers is not in the document]
Comment (leplatrem) [reply content/Dev/python_good_practices_2020.rst:106]: I mention pre-commit a bit below, and apparently https://github.com/pre-commit/mirrors-isort is deprecated

File: content/Dev/python_good_practices_2020.rst [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+:date: 2020-04-01
+
+
+I had the chance to spin off a new Python project recently, and that was a good opportunity to look at recent trends :) This article goes through some of the choices we made, knowing that almost everything is obviously debatable.
+
+
+Environment
+-----------
+
+Since we publish the app as a Docker container, we don't have to support multiple Python environments (eg. with `tox <https://tox.readthedocs.io>`_). Even if contributors may want to use `Pyenv <https://github.com/pyenv/pyenv>`_ to overcome the limitations of their operating system.
```
Comment (glasserc) [inline content/Dev/python_good_practices_2020.rst:14]: `Even` is a conjunction -- it should either be part of the previous sentence or we should reframe the sentence (something like `On the other hand, contributors...`)

File: content/Dev/python_good_practices_2020.rst [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+Environment
+-----------
+
+Since we publish the app as a Docker container, we don't have to support multiple Python environments (eg. with `tox <https://tox.readthedocs.io>`_). Even if contributors may want to use `Pyenv <https://github.com/pyenv/pyenv>`_ to overcome the limitations of their operating system.  
+
+I like to keep tooling minimalist, and I can't explain why, but I enjoy keeping the list of configuration files in the project root folder. 
+
+I'm familiar with ``make``, and it's quite universal and popular. So using a single ``Makefile`` with the appropriate dependencies between targets, we can create the environment and run the application or the tests by running one make command.
+
+I was used to Virtualenv, Pip, and requirements files. Common practice consists in having a folder with a requirements file by environment, and a constraint file for reproducible builds. We also setup `Dependabot <https://app.dependabot.com/>`_ to make sure our dependencies are kept up to date in the repo.
```
Comment (glasserc) [inline content/Dev/python_good_practices_2020.rst:20]: I think we could link to the folder too.

File: content/Dev/python_good_practices_2020.rst [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+
+- `flake8-docstrings <https://pypi.org/project/flake8-docstrings/>`_ or `darglint <https://github.com/terrencepreilly/darglint>`_ to validate your docstrings
+- `wemake-python-styleguide <https://github.com/wemake-services/wemake-python-styleguide#what-we-are-about>`_ for a very strict Python linter
+- `bandit <https://bandit.readthedocs.io/en/latest/>`_ to find common security issues
+
+
+Tests
+-----
+
+There's almost no debate about `pytest <https://pytest.readthedocs.io>`_. To me, the most appealing feature is the `fixtures decorator <https://docs.pytest.org/en/latest/fixture.html>`_, to maintain your tests `DRY <https://en.wikipedia.org/wiki/Don%27t_repeat_yourself>`_. It allows to obtain dependency injection, object factories, connection setup, config changes...
```
Comment (glasserc) [inline content/Dev/python_good_practices_2020.rst:145]: I would say `to keep your tests DRY`. Also, I wouldn't normally flag this, but I would not say `It allows [infinitive]` but rather `It allows you [infinitive]` or `It allows [gerund]`, and I wouldn't use the verb `obtain` for dependency injection -- maybe `use` or `perform` instead. Maybe rephrase as `It enables you to use dependency injection, object factories, connection setup, config changes...`

File: content/Dev/python_good_practices_2020.rst [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+
+    import sys
+
+    from poucave.app import main
+
+    main(sys.argv[1:])
+
+The most appreciated libraries for advanced CLI parameters seem to be `Click <https://click.palletsprojects.com>`_ (declarative) and `Fire <https://github.com/google/python-fire>`_ (automatic).
+
+For the Docker container, we follow our `Dockerflow conventions <https://github.com/mozilla-services/Dockerflow>`_. This helps our OPs team to treat all containers the same way, regardless of implementation language etc.
```
[the remark this answers is not in the document]
Comment (glasserc) [reply content/Dev/python_good_practices_2020.rst:214]: I generally think of the short form of "operations" as `ops` or `Ops`, and when I see `OPs` my brain tries to expand it as `Original Posts`. But maybe you're thinking of the IRC command `OP`?
